xcube/util/timeindex.py

```diff
@@ -0,0 +1,142 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+"""Utilities to ensure compatibility between time variables and their indexers
+
+The utilities in this module check and, where necessary, modify time indexers
+to ensure that they are compatible with the variables they are indexing.
+"""
+
+from typing import Dict, Any, Union
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+import logging
+
+logger = logging.getLogger('xcube')
+
+
+def ensure_time_label_compatible(var: Union[xr.DataArray, xr.Dataset],
+                                 labels: Dict[str, Any]) -> Dict[str, Any]:
+    """Ensure that labels['time'] is compatible with var
+
+    This function returns either the passed-in labels object, or a copy of
+    it with a modified value for labels['time'].
+
+    If there is no 'time' key in the labels dictionary or if there is no
+    'time' dimension in the var array, the original labels are returned.
+
+    If there is a 'time' key in the labels dictionary and a 'time' dimension
+    in the var array, they are checked for compatibility. If they are
+    compatible, the original labels are returned. If not, an altered labels
+    dictionary is returned, in which the time key has been modified to be
+    compatible with the type of the 'time' dimension in the var array.
+
+    See the documentation for `ensure_time_index_compatible` for details on
+    the compatibility check.
+    """
+
+    if 'time' in labels and 'time' in var.dims:
+        return dict(labels,
+                    time=ensure_time_index_compatible(var, labels['time']))
+    else:
+        return labels
+
+
+def ensure_time_index_compatible(var: Union[xr.DataArray, xr.Dataset],
+                                 timeval: Any) -> Any:
+    """Ensure that timeval is a valid time indexer for var
+
+    It is expected that the value of timeval will be a valid timestamp (i.e.
+    a valid input to pandas.Timestamp.__init__), or a slice in which the
+    start and stop fields are valid timestamps. For a slice, the start and
+    stop fields are processed separately, and their modified values (if
+    required) are returned as the start and stop fields of a new slice. The
+    step field is included unchanged in the new slice.
+
+    The compatibility check consists of checking the timezone-awareness
+    of the variable and its indexer. If the variable is timezone-aware and
+    the indexer is timezone-naive, or vice versa, a new indexer is returned
+    whose timezone-awareness corresponds to that of the variable. Otherwise
+    the original indexer is returned.
+    """
+    
+    if isinstance(timeval, slice):
```

Comment:
```suggestion
    timeval = labels[time_name]
    if isinstance(timeval, slice):
```

and so forth.

Reply:
I don't understand this suggestion. In this scope, `timeval` is already defined and `labels` is not bound, so we can't reassign it like this.